openvpn/build/build.py

```diff
@@ -152,131 +217,481 @@ def on_error(func, path, exc_info):
+        if '--owner=' in tar_options:
+            cmd.append('--owner=root')
+
+        cmd.append('-f')
+        cmd.append(os.path.join(
+            IMAGE_DIR, '{tarball_name}-staged.tar.gz'.format(
+                tarball_name=tarball_name)))
+        cmd.append('.')
+        run_command(cmd)
+
+
+def build_oqs_openssl(test_build=False):
+    '''
+    Build oqs_openssl
+    '''
+    if platform.system() == 'Windows':
+        build_oqs_openssl_windows(test_build)
+
+    if platform.system() == 'Linux' or platform.system() == 'Darwin':
+        build_oqs_openssl_unix(test_build)
+
+
+def build_openvpn(test_build=False):
+    '''
+    Build oqs_openssl
+    '''
+    if platform.system() == 'Windows':
+        build_openvpn_windows()
+
+    if platform.system() == 'Linux' or platform.system() == 'Darwin':
+        build_openvpn_unix()
+
+
+def pack_images():
+    '''
+    Build oqs_openssl
+    '''
+    if platform.system() == 'Windows':
+        pack_images_windows()
+
+    if platform.system() == 'Linux' or platform.system() == 'Darwin':
+        pack_images_unix()
+
+
+def post_messages():
+    '''
+    Display messages when the build has completed
+    '''
+    if platform.system() == 'Linux':
+        print(LINUX_UBUNTU_POST)
+
+
+#
+# commandline options
+#
+# enables --help
+
+
+parser = argparse.ArgumentParser(
+    description='Build OpenVPN PQ.',
+    epilog='Copyright 2018 Microsoft')
+
+parser.add_argument(
+    '--no-clean', action='store_true',
+    help=('clean build directory before.'),
+    default=False)
+parser.add_argument(
+    '--verbose', action='store_true',
+    help=('Be verbose.'),
+    default=False)
+
+parser.add_argument(
+    '--prefix', nargs='?', action='store', dest='prefix',
+    help=('prefix for the packaged results.'),
+    default=PREFIX)
+parser.add_argument(
+    '--build-dir', nargs='?', action='store', dest='build_dir',
+    help=('directory where to build.'),
+    default=BUILD_DIR)
+parser.add_argument(
+    '--sources-dir', nargs='?', action='store', dest='sources_dir',
+    help=('directory where to find downloaded content.'),
+    default=SOURCES_DIR)
+parser.add_argument(
+    '--image-dir', nargs='?', action='store', dest='image_dir',
+    help=('directory where to store the build results.'),
+    default=IMAGE_DIR)
+
+parser.add_argument(
+    '--skip-test', action='store_true',
+    help=('skip running make test on build binaries.'),
+    default=False)
+parser.add_argument(
+    '--skip-openssl', action='store_true',
+    help=('skip building openssl.'),
+    default=False)
+parser.add_argument(
+    '--skip-openvpn', action='store_true',
+    help=('skip building openvpn.'),
+    default=False)
+parser.add_argument(
+    '--skip-images', action='store_true',
+    help=('skip building images from staged.'),
+    default=False)
+parser.add_argument(
+    '--skip-download', action='store_true',
+    help=('skip downloading files, use BUILD_DIR contents.'),
+    default=False)
```

Comment:
So right now we have two scripts: build.py, which pulls from official sources and builds, and devbuild.py, which uses existing repos to allow developers to test changes on the fly. They're mostly copies of one another with some minor changes.

It looks like with --skip-download it may be possible for a developer to clone their own repos into the correct locations, make local changes, and build with them. Is this correct? If so, I think we can also delete devbuild.py as part of this change.

Reply:
No, its a build option that allows separating the download (network intensive) and the build (cpu-intensive) the --no-clean option may be used for development works, but needs more enhancement in git_restore (eg not unpacking when options.no_clean)
My opinion too is that a separate devbuild.py is overkill, should all be handled in single script